src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -82,6 +69,46 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
             return newTable.ToImmutableAndFree();
         }
 
+        private (ImmutableArray<TInput>, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>) GetValuesAndInputs(
+            NodeStateTable<TInput> sourceTable,
+            NodeStateTable<ImmutableArray<TInput>> previousTable,
+            NodeStateTable<ImmutableArray<TInput>>.Builder newTable)
+        {
+            var sourceValuesBuilder = BuilderAndStatistics<TInput>.Allocate();
```

Comment:
that's not the only goal.  a strong goal is to prevent wastage even when we need to generate new arrays.  in my testing we're continually creating temp arrays and purging them in all these cases due to the 128 item limit on normal pooling.  

i could triviall get around that by *always* pooling, but that was greatly worrying for me.  Imagine a case where for some reason we need to have a scratch array of 100k items at one point, but from taht point on, all future scratch arrays are like 100 elements.  We'd hold onto those 100k elements forever.  This approach allows that large allocation to decay out and eventually be recovered.

However, if we need a scratch array of 100k elements, and we're constantly producing around that many items on each incremental pass, then we def want to preserve that scratch buffer so that we don't constantly throw it away, and regen it, producing massive amounts of garbage and GC churn (note: these ar enot hypothetical concerns.  my other PRs are about directly addressing that being one of hte highest perf hits here) :)

Reply:
> a strong goal is to prevent wastage even when we need to generate new arrays. 

It looks like `ArrayBuilder.ToImmutableAndFree()` will use the internal array as-is if `_builder.Capacity == Count`. So for new arrays that exceed the 128 item limit, aren't we already allocating the minimal number of arrays (1 per `ArrayBuilder` / `ImmutableArray`)?

Reply:
It may be teh case that we can determine the precise length of this array up front for the batch case (and i can look into making tha tchange here).  But it's not a generally available facility for anythnig else (for example, any 'where' nodes will produce a different number of items).  :)

Reply:
Looking into simplifying this PR right now.

Reply:
this simplification has been done.